keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md

```diff
@@ -0,0 +1,891 @@
+  ability for controllers to proceed and not block.
+
+### User Stories (Optional)
+
+#### Garbage Collector
+
+The garbage collector makes decisions about deleting objects when all
+referencing objects are deleted. A discovery gap / apiserver mismatch, as
+described above, could result in GC seeing a 404 and assuming an object has been
+deleted; this could result in it deleting a subsequent object that it should
+not.
+
+This proposal will cause the GC to see the complete list of resources in
+discovery, and when it requests specific objects, see either the correct object
+or get a 503 (which it handles safely).
+
+#### Namespace Lifecycle Controller
+
+This controller seeks to empty all objects from a namespace when it is deleted.
+Discovery failures cause NLC to be unable to tell if objects of a given resource
+are present in a namespace. It fails safe, meaning it refuses to delete the
+namespace until it can verify it is empty: this causes slowness deleteing
+namespaces that is a common source of complaint.
+
+Additionally, if the NLC knows about a resource that the apiserver it is talking
+to does not, it may incorrectly get a 404, assume a collection is empty, and
+delete the namespace too early, leaving garbage behind in etcd. This is a
+correctness problem, the garbage will reappear if a namespace of the same name
+is recreated.
+
+This proposal addresses both problems.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+Cluster admins might not read the release notes and realize they should enable
+network/firewall connectivity between apiservers. In this case clients will
+receive 503s instead of transparently being proxied. 503 is still safer than
+today's behavior.
+
+Requests will consume egress bandwidth for 2 apiservers when proxied. We can cap
+the number if needed, but upgrades aren't that frequent and few resources are
+changed on releases, so these requests should not be common. We will count them
+with a metric.
+
+There could be a large volume of requests for a specific resource which might result in the identified apiserver being unable to serve the proxied requests. This scenario should not occur too frequently, since resource types which have large request volume should not be added or removed during an upgrade -- that would cause other problems, too.
+
+We should ensure at most one proxy, rather than proxying the request over and over again (if the source apiserver has an incorrect understanding of what the destination apiserver can serve).
+
+To prevent server-side request forgeries we will not give control over information about apiserver IP/endpoint and the trust bundle (used to authenticate server while proxying) to users via REST APIs.
```

Comment:
Can you describe the difference over endpoint control for this over say kubernetes.default.svc or aggregated APIs?  The trust bundle difference is fairly clear (the current in-cluster representation in kube-system is provided by the kube-apiserver, but not honored by it), but we already allow control of endpoints where the kube-apiserver acts as a front proxy.  For other cases, we have accepted the combination of

1. cluster-admin can configure network proxy and it cannot be reconfigured via API
2. cluster-admin can configure the trust bundle used to verify a kube-apiserver and it cannot be configured via API

Reply:
Summarized the reasons and details of using the masterlease reconciler to store/retrieve network location for kube-apiservers here https://github.com/kubernetes/enhancements/blob/46c73f6f8a0c5c5e322c353b6ba0a948725a0ad8/keps/sig-api-machinery/3903-unknown-version-interoperability-proxy/README.md#identifying-destination-apiservers-network-location